File: python3/reverse_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def print(self):
        it = self
        while it:
            print(it.val)
            it = it.next


def createList(l: list) -> ListNode:
    if not l:
        return None

    head = ListNode(l[0])
    it = head

    for ele in l[1:]:
        it.next = ListNode(ele)
        it = it.next

    return head


def reverseList(head: ListNode) -> ListNode:
    '''
    先存下所有的 node address, 之後一起做 reverse
    '''
    if head is None:
        return None

    ptrs = []
    it = head
    while it:
        ptrs.append(it)
        it = it.next

    for i in range(len(ptrs)):
        ptrs[i].next = ptrs[i-1]
    ptrs[0].next = None

    return ptrs[-1]


if __name__ == '__main__':
    # case 1
    head = createList([1, 2, 3, 4, 5, 6])
    result = reverseList(head)
    result.print()

    # case 2
    print('=====================')
    head = createList([1])
    result = reverseList(head)
    result.print()

    # case 3
    print('=====================')
    head = createList([1, 2])
    result = reverseList(head)
    result.print()

    # case 4
    print('=====================')
    head = createList([])
    result = reverseList(head)
    assert result is None

File: cpp/reverse_linked_list.cpp
#include <vector>
#include <iostream>
using namespace std;

class ListNode{
    public:
        int val;
        ListNode *next;
        ListNode(int x): val(x), next(NULL) {}
};


ListNode* reverseList(ListNode* head){
    if(head == NULL) return NULL;

    vector<ListNode*> ptrs;
    ListNode* it = head;
    while(it){
        ptrs.push_back(it);
        it = it->next;
    }

    for(unsigned int i=1;i<ptrs.size(); i++){
        ptrs[i]->next = ptrs[i-1];
    }

    ptrs[0]->next = NULL;

    return ptrs.back();
}


int main(){
    ListNode node1 = ListNode(1);
    ListNode node2 = ListNode(2);
    node1.next = &node2;
    ListNode* head = &node1;

    head = reverseList(head);
}
